server/id/reusable_id_impl.go

```diff
@@ -0,0 +1,55 @@
+package id
+
+import (
+	"context"
+	"sync"
+)
+
+type ReusableAllocatorImpl struct {
+	// RWMutex is used to protect following fields.
+	lock sync.Mutex
+
+	idQueue     []uint64
+	maxShardNum uint64
+}
+
+func NewReusableAllocatorImpl(maxShardNum uint64) *ReusableAllocatorImpl {
+	// Init id queue
+	var queue []uint64
+	for i := uint64(0); i < maxShardNum; i++ {
+		queue = append(queue, i)
+	}
+
+	return &ReusableAllocatorImpl{
+		idQueue:     queue,
+		maxShardNum: maxShardNum,
+	}
+}
+
+func (a *ReusableAllocatorImpl) isExhausted() bool {
+	return len(a.idQueue) == 0
+}
+
+func (a *ReusableAllocatorImpl) Collect(id uint64) error {
```

Comment:
Collect shardid in `NewReusableAllocatorImpl`?
Add `shardids` in `NewReusableAllocatorImpl` parameter?